docs/designs/sqlflow_integration_design.md

```diff
@@ -0,0 +1,237 @@
+    engine.worker_resource_request = "cpu=400m,memory=2048Mi",
+    engine.worker_resource_limit = "cpu=1,memory=3072Mi",
+    engine.num_workers = 2,
+    engine.volume = "",
+    engine.image_pull_policy = "Never",
+    engine.restart_policy = "Never",
+    engine.extra_pypi_index = "",
+    engine.namespace = "default",
+    engine.minibatch_size = 64,
+    engine.master_pod_priority = "",
+    engine.cluster_spec = "",
+    engine.num_minibatches_per_task = 2,
+    engine.docker_image_repository = "",
+    engine.envs = ""
+USING trained_elasticdl_keras_classifier;
+```
+
+
+### Implementation
+
+#### Mapping Extended SQL
+
+The components of the extended SQL defined by SQLFlow are mapped to a ``elasticDLFiller`` struct that looks like the following:
+
+```
+type elasticDLFiller struct {
+	IsTraining bool
+	TrainInputTable    string
+	EvalInputTable     string
+	PredictInputTable  string
+	PredictOutputTable string
+	PredictInputModel  string
+	OutputShape        int
+	InputShape         int
+	ModelDir           string
+	LabelColName        string
+	FeaturesList        string
+	TrainClause   *resolvedTrainClause
+	PredictClause *resolvedPredictClause
+}
+```
+
+This ``elasticDLFiller`` struct will be used to fill a template pre-defined to generate the model definition components required
+for ElasticDL, such as the model definition using `tf.keras` API, loss, optimizer, `dataset_fn`, etc.
+
+For example, the `dataset_fn` is generated using the `FeaturesList`, `LabelColName`, `InputShape`, `IsTraining`, and `TrainClause` in the ``elasticDLFiller`` struct:
+
+```python
+def dataset_fn(dataset, mode, metadata):
+    def _parse_data(record):
+
+        def _get_features_without_labels(
+            record, label_col_ind, features_shape
+        ):
+            features = [
+                record[:label_col_ind],
+                record[label_col_ind + 1 :],
+            ]
+            features = tf.concat(features, -1)
+            return tf.reshape(features, features_shape)
+
+        record = tf.strings.to_number(record, tf.float32)
+        features_shape = ({{.InputShape}}, 1)
+        labels_shape = (1,)
+        {{if .IsTraining}}
+        label_col_name = "{{.LabelColName}}"
+        if mode != Mode.PREDICTION:
+            if label_col_name not in metadata.column_names:
+                raise ValueError(
+                    "Missing the label column '%s' in the retrieved "
+                    "table." % label_col_name
+                )
+            label_col_ind = metadata.column_names.index(label_col_name)
+            labels = tf.reshape(record[label_col_ind], labels_shape)
+            return (
+                _get_features_without_labels(
+                    record, label_col_ind, features_shape
+                ),
+                labels,
+            )
+        {{end}}
+        return tf.reshape(record, features_shape)
+
+    dataset = dataset.map(_parse_data)
+
+    {{if .IsTraining}}
```

Comment:
Let us remove Go template parameters from the Python code. Instead, let us try passing in `{{.IsTraining}}` as a Python function parameter of `dataset_fn` or `_parse_data`. In this way, most part of the submitter template would be runnable Python code, which could have corresponding unit tests.

Ultimately, the template should be extremely simple -- a single funtion call, which parameters are template paramters. This function call in turn calls other functions, like `dataset_fn`.

Reply:
Yes, I agree that we should make the generated code as concise as possible and it would be a good next step. This doc is a summary of the current design, implementation, and brief outline of next steps.